File: Python Programming/tic-tac-toe.py
import numpy as np
import pandas as pd
import random
import time


def add_symbol(board, row, column, symbol):
    board.at[row, column] = symbol


def check_for_winner(board, symbol):
    board_check = pd.DataFrame(board == symbol)
    if board_check.all(axis='columns').any():
        print(symbol + " wins!")
        quit()

    if board_check.all(axis='index').any():
        print(symbol + " wins!")
        quit()

    if np.diag(board_check).all():
        print(symbol + " wins!")
        quit()

    if np.diag(np.fliplr(board_check)).all():
        print(symbol + " wins!")
        quit()

    if not available_fields:
        print("Tie!")
        quit()

    return


def player_move():
    while True:
        input_row = input("Row? ")
        if input_row == "quit":
            quit()
        elif input_row == "0" or input_row == "1" or input_row == "2":
            input_row = int(input_row)
        else:
            print("Invalid entry. Only enter 0, 1, 2, or quit")
            continue

        input_column = input("Column? ")
        if input_column == "quit":
            quit()
        elif input_column == "0" or input_column == "1" or input_column == "2":
            input_column = int(input_column)
        else:
            print("Invalid entry. Only enter 0, 1, 2, or quit")
            continue

        if (input_row, input_column) in available_fields:
            available_fields.remove((input_row, input_column))
            add_symbol(current_board, input_row, input_column, user_symbol)
            break
        else:
            print("Sorry, this field has already been used. Try again!")
            continue
    return


def pc_move():
    current_field = available_fields.pop(random.randint(0, len(available_fields) - 1))
    add_symbol(current_board, current_field[0], current_field[1], pc_symbol)
    return


user_symbol = None
pc_symbol = None
available_fields = [(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
current_board = pd.DataFrame({0: [".", ".", "."], 1: [".", ".", "."], 2: [".", ".", "."], })

while True:
    user_symbol = input("X or O? Type 'quit' at any prompt to exit. ")

    if user_symbol == "quit":
        quit()
    elif user_symbol == "X" or user_symbol == "x":
        user_symbol = "X"
        pc_symbol = "O"
        break
    elif user_symbol == "O" or user_symbol == "o":
        user_symbol = "O"
        pc_symbol = "X"
        break
    else:
        print("Invalid entry. Only enter X, O, or quit.")
        continue

for current_round in range(9):
    print("########################\n#####   ROUND #" + str(current_round + 1) + "   #####\n########################")
    time.sleep(2)
    if user_symbol == "X":
        # Player move
        print("CURRENT BOARD:\n------------------------")
        print(current_board)
        print("------------------------\n")
        print("Player X move:")
        player_move()
        print("CURRENT BOARD:\n------------------------")
        print(current_board)
        print("------------------------\n")
        time.sleep(2)
        check_for_winner(current_board, user_symbol)

        # PC move
        print("Player O move:")
        time.sleep(1)
        pc_move()
        print("Computer move registered.")
        time.sleep(2)
        print("CURRENT BOARD:\n------------------------")
        print(current_board)
        print("------------------------\n")
        time.sleep(2)
        check_for_winner(current_board, pc_symbol)
    else:
        # PC move
        print("CURRENT BOARD:\n------------------------")
        print(current_board)
        print("------------------------\n")
        print("Player X move:")
        time.sleep(1)
        pc_move()
        print("Computer move registered.")
        time.sleep(2)
        print("CURRENT BOARD:\n------------------------")
        print(current_board)
        print("------------------------\n")
        time.sleep(2)
        check_for_winner(current_board, pc_symbol)

        # Player move
        print("Player O move:")
        player_move()
        print("CURRENT BOARD:\n------------------------")
        print(current_board)
        print("------------------------\n")
        time.sleep(2)
        check_for_winner(current_board, user_symbol)

File: R Programming/tic-tac-toe.R
# Function definitions
player_move <- function(){
  while(TRUE) {
    cat("Row? ")
    input_row <-  readLines(con = con, n = 1)
    if(input_row == "quit") {
      quit()
    } else if(input_row == "1" | input_row == "2" | input_row == "3") {
      input_row <- as.numeric(input_row)
    } else {
      cat("Invalid input. Please only enter 1, 2, 3, or quit \n")
      next
    }
      
    cat("Column? ")
    input_column = readLines(con = con, n = 1)
    if(input_column == "quit") {
      quit()
    } else if(input_column == "1" | input_column == "2" | input_column == "3") {
      input_column <- as.numeric(input_column)
    } else {
      cat("Invalid input. Please only enter 1, 2, 3, or quit \n")
      next
    }
    
    if(any(sapply(available_fields, function(x) {all(x == c(input_row, input_column))}))) {
      available_fields[which(sapply(available_fields, function(x) {all(x == c(input_row, input_column))}))] <<- NULL
      current_board[input_row, input_column] <<- user_symbol
      break
    } else {
      cat("This field has already been filled. Please select an empty field. \n")
      next
    }
  }
}

pc_move <- function(){
  current_index <- sample.int(length(available_fields), 1)
  current_field <- available_fields[[current_index]]
  available_fields[[current_index]] <<- NULL
  current_board[current_field[1], current_field[2]] <<- pc_symbol
}

check_for_winner <- function(symbol){
  row_win = any(apply(current_board, 1, function(x) {isTRUE(all.equal(x, c(symbol, symbol, symbol)))}))
  column_win = any(apply(current_board, 2, function(x) {isTRUE(all.equal(x, c(symbol, symbol, symbol)))}))
  diagonal_win = isTRUE(all.equal(diag(current_board), c(symbol, symbol, symbol)))
  antidiagonal_win = isTRUE(all.equal(diag(current_board[nrow(current_board):1, ]), c(symbol, symbol, symbol)))
  
  if(row_win | column_win | diagonal_win | antidiagonal_win) {
    cat(paste0(symbol, " wins!"))
    quit()
  }
  if(length(available_fields) == 0) {
    cat("Tie!")
    quit()
  }
}

# Initialize script parameters
if (interactive()) {
  con <- stdin()
} else {
  con <- "stdin"
}

# Initialize game variables
user_symbol <-  NULL
pc_symbol <-  NULL
current_board <-  matrix(rep(NA, times = 9), nrow = 3, ncol = 3)
available_fields <-  list(c(1,1), c(1,2), c(1,3), c(2,1), c(2,2), c(2,3), c(3,1), c(3,2), c(3,3))


# Initial game setup
while(TRUE) {
  cat("X or O? ")
  user_symbol <- readLines(con = con, n = 1)
  
  if(user_symbol == "quit") {
    quit()
  } else if(user_symbol == "X" | user_symbol == "x") {
    user_symbol <- "X"
    pc_symbol <- "O"
    break
  } else if(user_symbol == "O" | user_symbol == "o") {
    user_symbol <- "O"
    pc_symbol <- "X"
    break
  } else {
    cat("Invalid entry. Please only enter X, O, or quit \n")
    next
  }
}

#Game
for (current_round in 1:9) {
  cat(paste0("################# Round ", current_round, ": #################\n"))
  Sys.sleep(1)
  cat("Current board:\n")
  print(current_board)
  cat("\n\n")
  if(user_symbol == "X"){
    # Player move
    cat("Player X move\n")
    Sys.sleep(1)
    player_move()
    Sys.sleep(1)
    cat("Current board:\n")
    print(current_board)
    cat("\n\n")
    check_for_winner(user_symbol)
    Sys.sleep(1)
    
    # PC move
    cat("Player O move\n")
    Sys.sleep(1)
    cat("Executing computer move...\n")
    pc_move()
    Sys.sleep(1)
    cat("Current board:\n")
    print(current_board)
    cat("\n\n")
    check_for_winner(pc_symbol)
    Sys.sleep(1)
  } else {
    # PC move
    cat("Player X move\n")
    Sys.sleep(1)
    cat("Executing computer move...\n")
    pc_move()
    Sys.sleep(1)
    cat("Current board:\n")
    print(current_board)
    cat("\n\n")
    check_for_winner(pc_symbol)
    Sys.sleep(1)
    
    # Player move
    cat("Player O move\n")
    Sys.sleep(1)
    player_move()
    Sys.sleep(1)
    cat("Current board:\n")
    print(current_board)
    cat("\n\n")
    check_for_winner(user_symbol)
    Sys.sleep(1)
  }
}